### PR TITLE
Control the maximum amount of events in single logRequest to be 1000 in Performance Monitoring

### PR DESCRIPTION
**Pull request by zijianjoy**
Backend starts rejecting logRequest with more than 1000 events in batch. In order for client event to be accepted and avoid any unintentional flood, put a cap for each logRequest dispatch to be 1000, extra events will stay in queue until next dispatch.

### CODE REVIEW

**Review by jposuna — APPROVED** [review]

File: packages-exp/performance-exp/src/services/transport_service.ts [hunk 1]
```diff
@@ -90,9 +91,11 @@ function processQueue(timeOffset: number): void {
 }
 
 function dispatchQueueEvents(): void {
-  // Capture a snapshot of the queue and empty the "official queue".
-  const staged = [...queue];
-  queue = [];
+  // Extract events up to the maximum cap of single logRequest from top of "official queue".
+  // The staged events will be used for current logRequest attempt, remaining events will be kept
+  // for next attempt.
+  const staged = queue.slice(0, MAX_EVENT_COUNT_PER_REQUEST);
```
Comment (alikn) [inline packages-exp/performance-exp/src/services/transport_service.ts:97]: use splice instead.

File: packages/performance/src/services/transport_service.test.ts [hunk 2]
```diff
@@ -78,36 +80,128 @@ describe('Firebase Performance > transport_service', () => {
     clock.tick(INITIAL_SEND_TIME_DELAY_MS);
     testTransportHandler('someEvent');
     clock.tick(DEFAULT_SEND_INTERVAL_MS);
+    await Promise.resolve();
```
Comment (alikn) [inline packages/performance/src/services/transport_service.test.ts:83]: Why do we need this?

File: packages/performance/src/services/transport_service.test.ts [hunk 3]
```diff
@@ -78,36 +80,128 @@ describe('Firebase Performance > transport_service', () => {
...
-      )
+      generateSuccessResponse()
     );
 
     testTransportHandler('event1');
     clock.tick(INITIAL_SEND_TIME_DELAY_MS);
     expect(fetchStub).to.have.been.calledOnce;
   });
+
+  it('sends up to the maximum event limit in one request', async () => {
```
Comment (alikn) [inline packages/performance/src/services/transport_service.test.ts:101]: This unit test is hard to follow. Can we instead have one in which there is more than 1000 logs in the queue and check the number of logs received by fetch stub to be equal to 1000?

File: packages-exp/performance-exp/src/services/transport_service.ts (hunk 1 above)
Comment (zijianjoy) [reply packages-exp/performance-exp/src/services/transport_service.ts]: Good suggestion! Applied `splice()` to split the array.

File: packages/performance/src/services/transport_service.test.ts (hunk 2 above)
Comment (zijianjoy) [reply packages/performance/src/services/transport_service.test.ts]: Removed.

File: packages/performance/src/services/transport_service.test.ts (hunk 3 above)
Comment (zijianjoy) [reply packages/performance/src/services/transport_service.test.ts:101]: Simplified unit test to capture only the first dispatch to be equal to 1000 events. Since we also have the splice logic of keeping the remaining events in the queue, added validation on remaining event count is kept.

File: packages-exp/performance-exp/src/services/transport_service.test.ts [hunk 4]
```diff
@@ -82,32 +84,102 @@ describe('Firebase Performance > transport_service', () => {
...
+    // Returns successful response from fl for logRequests.
+    const response = generateSuccessResponse();
+    fetchStub
+      .withArgs(flTransportFullUrl, {
+        method: 'POST',
+        body: JSON.stringify(logRequest)
+      })
+      .resolves(response);
+    stub(response, 'json').returns(
+      Promise.resolve(JSON.parse(generateSuccessResponseBody()))
```
Comment (alikn) [inline packages-exp/performance-exp/src/services/transport_service.test.ts:123]: Here we can use stub(...).resolve instead of stub(...).returns(Promise.resolve(...))

File: packages-exp/performance-exp/src/services/transport_service.test.ts [hunk 5]
```diff
@@ -82,32 +84,102 @@ describe('Firebase Performance > transport_service', () => {
...
+    // Assert
+    clock.tick(INITIAL_SEND_TIME_DELAY_MS);
+    // First logRequest has been called.
+    expect(fetchStub).to.have.been.calledWith(flTransportFullUrl, {
+      method: 'POST',
+      body: JSON.stringify(logRequest)
+    });
+    // Wait for async action to call for next dispatch cycle.
+    await Promise.resolve()
+      .then(() => {
```
Comment (alikn) [inline packages-exp/performance-exp/src/services/transport_service.test.ts:141]: I think this is equivalent to 
Promise.resolve()
.then(() => {
clock.tick(1);
clock.tick(1);
}

File: packages-exp/performance-exp/src/services/transport_service.test.ts [hunk 6]
```diff
@@ -82,32 +84,102 @@ describe('Firebase Performance > transport_service', () => {
...
+        clock.tick(1);
+      })
+      .then(() => {
+        clock.tick(1);
+      });
+    clock.tick(DEFAULT_SEND_INTERVAL_MS * 3);
+    // Remaining events stay in queue.
+    expect(readQueue().length).to.be.equal(20);
```
Comment (alikn) [inline packages-exp/performance-exp/src/services/transport_service.test.ts:149]: Shouldn't the 20 logs be sent in the next batch?

File: packages-exp/performance-exp/src/services/transport_service.ts [hunk 7]
```diff
@@ -65,6 +66,10 @@ export function setupTransportService(): void {
   }
 }
 
+export function readQueue(): BatchEvent[] {
```
Comment (alikn) [inline packages-exp/performance-exp/src/services/transport_service.ts:69]: Let's avoid exposing the internals of the service just for testing.

File: packages-exp/performance-exp/src/services/transport_service.test.ts (hunk 4 above)
Comment (zijianjoy) [reply packages-exp/performance-exp/src/services/transport_service.test.ts]: Updated, thank you Ali!

File: packages-exp/performance-exp/src/services/transport_service.test.ts (hunk 5 above)
Comment (zijianjoy) [reply packages-exp/performance-exp/src/services/transport_service.test.ts]: Thank you Ali! After investigation, I found that `clock.tick()` is not required in this promise chain, removing them.

File: packages-exp/performance-exp/src/services/transport_service.test.ts (hunk 6 above)
Comment (zijianjoy) [reply packages-exp/performance-exp/src/services/transport_service.test.ts]: That is correct, switched test logic to validate two batches with 1000 and 20 events each.

File: packages-exp/performance-exp/src/services/transport_service.ts (hunk 7 above)
Comment (zijianjoy) [reply packages-exp/performance-exp/src/services/transport_service.ts]: I agree, I removed this exposure, and validate the `fetch` calls with expected parameter instead.

**Review by alikn — APPROVED** [review]

**Review by hsubox76 — APPROVED** [review]
